File: admin-ui/store/sponsor.js
import Vue from "vue";

export const state = () => ({
  sponsors: [],
  sponsor: null,
});

export const getters = {
  sponsors(state) {
    return state.sponsors;
  },
  sponsor(state) {
    return state.sponsor;
  },
};

export const mutations = {
  SET_SPONSORS(state, sponsors) {
    Vue.set(state, "sponsors", sponsors);
  },
  SET_SPONSOR(state, sponsor) {
    Vue.set(state, "sponsor", sponsor);
  },
};

export const actions = {
  async create(
    _,
    {
      name,
      logo,
    },
  ) {
    const data = new FormData();
    data.set("name", name);
    data.set("logo", logo);

    const response = await this.$api.$post(
      "/sponsor",
      data,
    );

    return response;
  },

  async update(
    _,
    {
      id,
      name,
      logo,
    },
  ) {
    const data = new FormData();
    data.set("name", name);
    if (logo) {
      data.set("image", logo);
    }

    const response = await this.$api.$patch(
      `/sponsor/${ id }`,
      data,
    );

    return response;
  },

  async remove(
    {
      dispatch,
    },
    sponsor,
  ) {
    const response = await this.$api.$delete(
      `/sponsor/${ sponsor.id }`,
    );

    if ("success" === response?.status) {
      await dispatch("listAll");
    }

    return response;
  },

  async listAll(
    {
      commit,
    },
  ) {
    const response = await this.$api.$get(
      "/sponsor",
    );

    const { data = [] } = response || {};

    commit("SET_SPONSORS", data);

    return data;
  },

  async swap(
    {
      commit,
    },
    {
      a,
      b,
    },
  ) {
    const response = await this.$api.$patch(
      `/sponsor/swap/${ a.id }`,
      {
        with: b.id,
      },
    );

    if ("success" === response?.status) {
      commit("SET_SPONSORS", response.data);
    }

    return response;
  },

  async info(
    {
      commit,
    },
    {
      id,
    },
  ) {
    const response = await this.$api.$get(
      `/sponsor/${ id }`,
    );

    const { status = "error", data = null } = response || {};

    if ("success" === status) {
      commit("SET_SPONSOR", data);
    } else {
      commit("SET_SPONSOR", null);
    }

    return data;
  },
};

File: app/routes/api/sponsor/sponsor.go
package sponsor

import (
	"mime/multipart"

	"github.com/gofiber/fiber/v2"
	"github.com/google/uuid"
	"github.com/valyala/fasthttp"

	"brucosijada.kset.org/app/models"
	"brucosijada.kset.org/app/providers/database"
	"brucosijada.kset.org/app/repo"
	"brucosijada.kset.org/app/response"
)

func CreateSponsor(ctx *fiber.Ctx) (err error) {
	type form struct {
		Name string
	}

	var input form
	if err := ctx.BodyParser(&input); err != nil {
		return ctx.Status(fiber.StatusBadRequest).JSON(
			response.Error(
				"Invalid request",
				err,
			),
		)
	}

	var logo *multipart.FileHeader
	if logo, err = ctx.FormFile("logo"); err != nil {
		return ctx.Status(fiber.StatusBadRequest).JSON(
			response.Error(
				"Invalid request: Missing file",
				err,
			),
		)
	}

	sponsor, err := repo.Sponsor().Create(
		input.Name,
		logo,
		ctx.Locals("user").(models.User).ID,
	)

	if err != nil {
		return ctx.Status(fiber.StatusUnprocessableEntity).JSON(
			response.Error(
				err.Error(),
				err,
			),
		)
	}

	return ctx.JSON(
		response.Success(
			sponsor,
		),
	)
}

func ListSponsors(ctx *fiber.Ctx) error {
	items, err := repo.Sponsor().ListSimple()
	if err != nil {
		return ctx.Status(fiber.StatusInternalServerError).JSON(
			response.Error(
				"Something went wrong",
				err,
			),
		)
	}

	return ctx.JSON(
		response.Success(
			items,
		),
	)
}

func UpdateSponsor(ctx *fiber.Ctx) (err error) {
	type form struct {
		Name string
	}

	db := database.DatabaseProvider().Client()

	var sponsor = models.Sponsor{}
	db.Model(sponsor).Preload("Image").Preload("Image.Variations").Where(
		"uuid = ?",
		ctx.Params("id"),
	).First(&sponsor)

	if !sponsor.Exists() {
		return ctx.Status(fiber.StatusNotFound).JSON(
			response.Error(
				"Sponsor not found",
				nil,
			),
		)
	}

	var input form
	if err := ctx.BodyParser(&input); err != nil {
		return ctx.Status(fiber.StatusBadRequest).JSON(
			response.Error(
				"Invalid request",
				err,
			),
		)
	}

	var logo *multipart.FileHeader
	if logo, err = ctx.FormFile("image"); err != nil && err != fasthttp.ErrMissingFile {
		return ctx.Status(fiber.StatusBadRequest).JSON(
			response.Error(
				"Invalid request: Missing file",
				err,
			),
		)
	}

	err = repo.Sponsor().Update(
		&sponsor,
		input.Name,
		logo,
		ctx.Locals("user").(models.User).ID,
	)

	if err != nil {
		return ctx.Status(fiber.StatusUnprocessableEntity).JSON(
			response.Error(
				err.Error(),
				err,
			),
		)
	}

	return ctx.JSON(
		response.Success(
			repo.Sponsor().ToSponsorItem(&sponsor),
		),
	)
}

func ShowSponsor(ctx *fiber.Ctx) (err error) {
	var id uuid.UUID

	if id, err = uuid.Parse(ctx.Params("id")); err != nil {
		return ctx.Status(fiber.StatusUnprocessableEntity).JSON(
			response.Error(
				"Invalid sponsor ID",
				err,
			),
		)
	}

	sponsor, err := repo.Sponsor().GetByUuidSimple(id)
	if err != nil {
		return ctx.Status(fiber.StatusInternalServerError).JSON(
			response.Error(
				err.Error(),
				err,
			),
		)
	}

	return ctx.JSON(
		response.Success(
			sponsor,
		),
	)
}

func SwapSponsors(ctx *fiber.Ctx) (err error) {
	input := struct {
		With string `json:"with"`
	}{}
	if err := ctx.BodyParser(&input); err != nil {
		return ctx.Status(fiber.StatusBadRequest).JSON(
			response.Error(
				"Invalid request",
				err,
			),
		)
	}

	var swap, with uuid.UUID

	if swap, err = uuid.Parse(ctx.Params("id")); err != nil {
		return ctx.Status(fiber.StatusUnprocessableEntity).JSON(
			response.Error(
				"Invalid swap sponsor ID",
				err,
			),
		)
	}
	if with, err = uuid.Parse(input.With); err != nil {
		return ctx.Status(fiber.StatusUnprocessableEntity).JSON(
			response.Error(
				"Invalid with sponsor ID",
				err,
			),
		)
	}

	err = repo.Sponsor().Swap(swap, with)
	if err != nil {
		return ctx.Status(fiber.StatusInternalServerError).JSON(
			response.Error(
				"Something went wrong",
				err,
			),
		)
	}

	newSponsors, err := repo.Sponsor().ListSimple()
	if err != nil {
		return ctx.Status(fiber.StatusInternalServerError).JSON(
			response.Error(
				"Something went wrong",
				err,
			),
		)
	}

	return ctx.JSON(
		response.Success(
			newSponsors,
		),
	)
}

func DeleteSponsor(ctx *fiber.Ctx) (err error) {
	db := database.DatabaseProvider().Client()

	var sponsor models.Sponsor
	db.Model(&models.Sponsor{}).Preload("Image").Where("uuid = ?", ctx.Params("id")).First(&sponsor)

	if !sponsor.Exists() {
		return ctx.JSON(
			response.Success(
				nil,
			),
		)
	}

	if err = db.Delete(&sponsor).Error; err != nil {
		return ctx.JSON(
			response.Error(
				"Something went horribly wrong",
				err,
			),
		)
	}

	return ctx.JSON(
		response.Success(
			sponsor,
		),
	)
}
